src/Analyzers/CSharp/Analyzers/ConvertTypeofToNameof/CSharpConvertTypeofToNameofDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,108 @@
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Operations;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.CSharp.ConvertTypeofToNameof
+{
+    /// <summary>
+    /// Finds code like typeof(someType).Name and determines whether it can be changed to nameof(someType), if yes then it offers a diagnostic
+    /// </summary>
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    internal class CSharpConvertTypeofToNameofDiagnosticAnalyzer : AbstractBuiltInCodeStyleDiagnosticAnalyzer
+    {
+        public CSharpConvertTypeofToNameofDiagnosticAnalyzer()
+            : base(IDEDiagnosticIds.ConvertTypeOfToNameOfDiagnosticId,
+                   option: null,
+                   LanguageNames.CSharp,
+                   new LocalizableResourceString(
+                       nameof(CSharpAnalyzersResources.Convert_typeof_to_nameof), CSharpAnalyzersResources.ResourceManager, typeof(CSharpAnalyzersResources)))
+        {
+        }
+
+        protected override void InitializeWorker(AnalysisContext context)
+        {
+            context.RegisterOperationAction(AnalyzeTypeOfAction, OperationKind.TypeOf);
+        }
+
+        private void AnalyzeTypeOfAction(OperationAnalysisContext context)
+        {
+            var syntaxTree = context.Operation.Syntax.SyntaxTree;
+            var node = context.Operation.Syntax;
+
+            // Make sure that the syntax that we're looking at is actually a typeof expression
+            if (!(node is TypeOfExpressionSyntax))
+            {
+                return;
+            }
+
+            // nameof was added in CSharp 6.0, so don't offer it for any languages before that time
+            if (((CSharpParseOptions)syntaxTree.Options).LanguageVersion < LanguageVersion.CSharp6)
+            {
+                return;
+            }
+
+            // Make sure the parent syntax is a member access expression otherwise the syntax is not the
+            // kind of expression that we want to analyze
+            if (!(node.Parent is MemberAccessExpressionSyntax))
+            {
+                return;
+            }
+
+            // Check if the operation is one that we want to offer the fix for
+            if (!IsValidOperation(context.Operation))
+            {
+                return;
+            }
+
+            var parent = node.Parent;
+            // If the parent node is null then it cannot be a member access, so do not report a diagnostic
+            if (parent is null)
+            {
+                return;
+            }
+
+            // Current case can be effectively changed to a nameof instance so report a diagnostic
+            var location = parent.GetLocation();
+            context.ReportDiagnostic(DiagnosticHelper.Create(Descriptor, location, ReportDiagnostic.Hidden, additionalLocations: null,
+                properties: null, messageArgs: null));
```

Comment:
```suggestion
            context.ReportDiagnostic(DiagnosticHelper.Create(
                Descriptor, location, ReportDiagnostic.Hidden, additionalLocations: null, properties: null, messageArgs: null));
```

(prefer keeping parameters together if possible, versus arbitrarily breaking them somewhere in the middle)

Reply:
I believe we should just `Diagnostic.Create` here. `DiagnosticHelper.Create` is only really required if you are reporting a diagnostic with code style option, where the effective severity might come from code style option.

Reply:
interesting.  my ideal would still be to have everyone go through one entrypoint, that way it's easy to tell immediately if the analyzer is usign the proper pattern.